src/control/server/config/server.go

```diff
@@ -496,15 +456,86 @@ func (cfg *Server) Validate(ctx context.Context, log logging.Logger) (err error)
 		return FaultConfigEvenAccessPoints
 	}
 
-	for i, engine := range cfg.Engines {
-		engine.Fabric.Update(cfg.Fabric)
+	cfgHasBdevs := false
+	cfgTargetCount := 0
+	for idx, ec := range cfg.Engines {
+		cfgTargetCount += ec.TargetCount
 
-		if err := engine.Validate(ctx, log); err != nil {
-			return errors.Wrapf(err, "I/O Engine %d failed config validation", i)
+		ls := ec.LegacyStorage
+		if ls.WasDefined() {
+			log.Infof("engine %d: Legacy storage configuration detected. Please "+
+				"migrate to new-style storage configuration.", idx)
+			var tierCfgs storage.TierConfigs
+			if ls.ScmClass != storage.ClassNone {
+				tierCfgs = append(tierCfgs,
+					storage.NewTierConfig().
+						WithScmClass(ls.ScmClass.String()).
+						WithScmDeviceList(ls.ScmConfig.DeviceList...).
+						WithScmMountPoint(ls.MountPoint).
+						WithScmRamdiskSize(ls.RamdiskSize),
+				)
+			}
+
+			// Do not add bdev tier if cls is none or nvme has no devices to maintain
+			// backward compatible behavior.
+			bc := ls.BdevClass
+			switch {
+			case bc == storage.ClassNvme &&
+				len(ls.BdevConfig.DeviceList) == 0:
+				log.Debugf("legacy storage config conversion skipped for class "+
+					"%s with empty bdev_list", storage.ClassNvme)
+			case bc == storage.ClassNone:
+				log.Debugf("legacy storage config conversion skipped for class %s",
+					storage.ClassNone)
+			default:
+				tierCfgs = append(tierCfgs,
+					storage.NewTierConfig().
+						WithBdevClass(ls.BdevClass.String()).
+						WithBdevDeviceCount(ls.DeviceCount).
+						WithBdevDeviceList(ls.BdevConfig.DeviceList...).
+						WithBdevFileSize(ls.FileSize).
+						WithBdevBusidRange(
+							ls.BdevConfig.BusidRange.String()),
+				)
+			}
+			ec.WithStorage(tierCfgs...)
+			ec.LegacyStorage = engine.LegacyStorage{}
+		}
+
+		if ec.Storage.Tiers.CfgHasBdevs() {
+			cfgHasBdevs = true
+			if ec.TargetCount == 0 {
+				return errors.Errorf("engine %d: Target count cannot be zero if "+
+					"bdevs have been assigned in config", idx)
+			}
+		}
+
+		ec.Fabric.Update(cfg.Fabric)
+
+		if err := ec.Validate(log, fis); err != nil {
+			return errors.Wrapf(err, "I/O Engine %d failed config validation", idx)
+		}
+
+		log.Debugf("engine %d fabric numa %d, storage numa %d", idx,
+			ec.Fabric.NumaNodeIndex, ec.Storage.NumaNodeIndex)
+	}
+
+	if cfgHasBdevs {
+		// Calculate minimum number of hugepages for all configured engines.
+		minHugePages, err := common.CalcMinHugePages(hugePageSize, cfgTargetCount)
+		if err != nil {
+			return err
 		}
 
-		log.Debugf("engine %d fabric numa %d, storage numa %d", i,
-			engine.Fabric.NumaNodeIndex, engine.Storage.NumaNodeIndex)
+		// If the config doesn't specify hugepages, use the minimum. Otherwise, validate
+		// that the configured amount is sufficient.
+		if cfg.NrHugepages == -1 {
```

Comment:
Hmm. I know that we decided that `nr_hugepages: 0` needs to be a valid setting for a special use case (explicit config with no bdevs?), but I'm sort of inclined to use `-1` for that value so that we can take advantage of the zero value as "unset" in order to avoid subtle errors.

I think it will also make it easier to fix whatever code was introduced that made it a requirement for the admin to sometimes set an explicit "no hugepages, thanks" config parameter -- that's kind of terrible from a UX perspective.

Reply:
ok, will update

Reply:
done